### ISSUE TITLE
Define primitive Visuals and separate gloo interfaces

### ISSUE THREAD

**Issue by djhoese** [vispy/vispy#2000]
# Overview

It was discussed at the end of the February 2021 sprint that in order for simplify maintenance of VisPy the `gloo` interfaces (or OpenGL in general) needs to be separated more completely from the rest of VisPy; mainly the plotting, SceneCanvas, and Visuals interfaces. The end result of this should be a clear (hopefully) separation of high-level VisPy and low-level "graphics" systems. This should allow us to leverage other visualization technologies in the future, hopefully sooner rather than later (ex. datoviz). 

By defining a set of "primitive" Visual classes whose functionality is provided a low-level graphics library (or gloo) we hope that the rest of VisPy's Visuals can be created without getting too low-level. If all goes well this means that we can easily switch an application from OpenGL to Vulkan (via datoviz) or WGPU (via wgpu-py) with something as simple as a global configuration setting or backend choice.

# Path Forward

1. Define primitives
2. Identify and fix OpenGL "leaks" where low-level gloo interfaces are leaking into higher level interfaces. Anything that is specific to OpenGL needs to be moved to a more generic Visual interface. For example `self.set_gl_state(...)` should be `self.set_graphics_behavior(...)` or something like that. This also includes any Visuals that are doing things that only a primitive Visual should be doing (setting shader properties directly).
3. Create proof of concept example using datoviz backend.

# Primitive Definitions

I've started this Wiki document with the definitions based on what I see in current VisPy:

https://github.com/vispy/vispy/wiki/Primitive-Visuals

# How to implement it

My only idea so far: Wrapper Visual classes for primitive Visuals that communicate with the lower-level gloo/datoviz primitive Visuals. The primitive class that actually lives in other libraries would exist as an instance attribute in the wrapper. This may be needed no matter what, but it also makes it easier to choose the low-level library at runtime (instead of import time). And allows for better subclassing.

I also keep wanting to call this a "visual backend" but that is confusing with the Application backend...or GUI backend?

```
class MeshVisual(BasePrimitiveVisual):
    def __init__(self, ...):
        self._backend = vispy.get_visual_backend()
        if self._backend == ' datoviz':
            from datoviz import get_panel
            self._panel = get_panel(...)  # this function doesn't actually exist
            self._primitive = self._panel.visual('mesh')
        elif ...

    @property
    def color(self):
        return self._primitive.color

    @color.setter
    def color(self, new_color):
        self._primitive.color = new_color
        self.update()

```

# Related

* #1988

**Comment by djhoese** [comment]
Please update the wiki with more primitive information or suggest changes here.

**Comment by almarklein** [comment]
So IIUC the idea would be to create a visuals API that uses any of several backends (gloo, dataviz, wgpu/pygfx) and then deprecate gloo when the other backends are mature enough. Sounds like a good plan to me.

In terms of implementation, I don't think you'd want to triage inside the classes, but implement each visual-backend in a subpackage. So the wrapper visuals become a thin layer around what's inside those, and they add the docs for the public API. That's a lot of layer, so perhaps subclassing can be used somehow, as long as it does not make things overly complex.


>  Can 2D Visuals be used in 3D and vice versa?

In most cases It will. 2D is just 3D looking from above :)

> Does we need to separate GUI backends and visual "backends"? Or does it become 'pyqt_gl' and 'pyqt_datoviz' or 'pyqt_vulkan'?

Good question. I don't think it's possible to completely separate these two backends. At least not the part that's needed to let the underlying lib (dataviz or wgpu) draw on it. But other things (e.g. events) is only specific to the GUI toolkit.

**Comment by djhoese** [comment]
> In terms of implementation, I don't think you'd want to triage inside the classes, but implement each visual-backend in a subpackage. So the wrapper visuals become a thin layer around what's inside those, and they add the docs for the public API. That's a lot of layer, so perhaps subclassing can be used somehow, as long as it does not make things overly complex.

The main thing I want to avoid is complicating users' lives. That means (I think) a user script would look like this:

```
from vispy.visuals import MeshVisual, LineVisual, SphereVisual
```

where mesh and line are wrappers around the graphics library (that's what I'm going to call datoviz, gloo, and pygfx unless someone has a better suggestion) and the sphere is a wrapper around the MeshVisual. The user shouldn't need to know which visuals are primitives (ex. `from vispy.visuals.datoviz_primitives import MeshVisual`). They also shouldn't need to know what graphics backend is being used. Similar to how it is possible to use a VisPy Canvas and make a full visualization without even knowing what backend vispy defaulted to.

I think subclassing will be prone to bugs if you mean `class MeshVisual(VisPyBaseVisual, DatovizMeshVisual):` as there will likely be conflicting properties and method names. Since the VisPy implementations will be doing completely different things from the graphics library, we'd likely just break the low-level library. Unless you mean doing something like:

```
# NOTE: I don't want to do at-import stuff if we can help it but not sure where/how this happens
if some_global_state == 'datoviz':
    from vispy._graphics_backend.datoviz import DatovizMeshVisualWrapper as PrimitiveWrapper
elif some_global_state == 'wgpu':
    ...

class MeshVisual(PrimitiveWrapper):
    ...

```

where the wrapper is:

```
class DatovizMeshVisualWrapper(VisPyBaseVisual):
    def __init__(self, ...):
        self._primitive = datoviz.panel.visual('mesh')
```

> Good question. I don't think it's possible to completely separate these two backends. At least not the part that's needed to let the underlying lib (dataviz or wgpu) draw on it. But other things (e.g. events) is only specific to the GUI toolkit.

I think this would be helped if we could figure out #1995 so that the current gloo backends are class-based. Thinking about it more, the backends really control two things: the Application and the Canvas. I wonder if we could leverage this to have an "app_backend" and "canvas_backend"...except it seems events are all produced and handled by the canvas :thinking: Scratch this idea.

**Comment by rougier** [comment]
We could also think in terms of API we want instead of object properties. For example, for line, in an ideal world I would like to be able to write something like:

```
from vispy.visuals import Line
scene = Line(position = np.random.uniform(0, 1, (50,3)),
             color = lambda x, y, z, l, d: x,y,z,d,
             width = lambda x, y, z, l, depth: l)
scene.show()
```

The `color` could be a numpy array or a lambda function (only) that would take the x,y,z vertex position + the curvilinear coordinate (normalized from 0 to 1) and the depth of the vertex (normalized from 0 to 1). These two are only accessible from inside the shader and it would require the lambda function to be translated into shader code. For simple lamnda function, this could be done without any conversion.

**Comment by djhoese** [comment]
I have a feeling that would come down to the specifics of the low-level graphics library and what kind of shader manipulation could be done (ex. datoviz not currently allowing non-linear transforms). I think if you give people a callable argument, they're going to put non-trivial things in there and I would prefer if VisPy didn't have to maintain a python->GLSL converter (like @almarklein tried and regrets).

I think something like what you're suggesting would require a subclass where a custom shader could be provided (even if just the color portion). I like the simplicity of a user being able to pass a shader (in any form) to the `__init__` but that is also in conflict with the purpose of the Visuals which is to be simple and not require GL understanding.

In terms of a subclass doing something like this, I can imagine VisPy providing the ability to customize individual components of the overall shader and then joining of various shader components similar to what it does already. The low-level library doesn't know anything except that it is getting a different shader.

**Comment by rossant** [comment]
I would advocate for a simple system where shaders are mostly fixed. Dynamically-generated shaders require a much higher degree of complexity.

Eventually it all comes down to the functionality needed from a GPU visualization backend. What is the minimal set of features required? For example:

- specify a viewport in the canvas, in pixel coordinates
- specify the data coordinates of the viewport
- add a visual among a list of predefined visuals (marker, path, image, mesh...)
- set visual properties
- set a controller among a set of predefined controllers
- picking: convert screen coordinates to data coordinates, and return the object and object part under the mouse pointer

By agreeing on this common set of features, any high-level API could use any low-level graphics backend that implements it. Perhaps one could start writing a few test case examples, with a high-level API, and try to implement it on top of a hypothetical backend to see what features would be needed.

**Comment by djhoese** [comment]
@rossant To be clear, I'm considering shaders that get generated at-runtime be limited to VisPy's side of things if necessary. It'd be great to have static shaders all the time, but at some point I think we're going to need them in VisPy. The big feature that uses them that I can think of is transforms. VisPy currently chains transforms from different levels of the scene/node graph into one set of function calls. You mentioned that datoviz only supports MVP transforming right now, but do you think non-linear transforms would be possible in the future. If so, how might that work from a user point of view? How do more advanced scene graph implementations (i.e. not vispy) handle this logic of multiple levels of transforms? Or is it not offered as a feature?

Regarding dynamic shaders in general, I would hope that we'd never need to update the core shader code for a particular Visual. For example, a property change on the Visual should never cause a change in the structure of the shader code; only values of uniforms, buffers, etc.

I've included a "Shared Functionality" section in the wiki page as well as a "Canvas" section. I hope these answer some of your questions about what the minimal set of features are. I'll see if I can update it with some of the stuff you've mentioned.

**Comment by djhoese** [comment]
@rossant What did you mean by "specify the data coordinates of the viewport"?

**Comment by rougier** [comment]
On example of what could be done with more versatile shaders: https://github.com/unconed/shadergraph (look for the demo)
Some crude / partial python implementation: https://github.com/rougier/shadergraph

**Comment by rossant** [comment]
> @rossant What did you mean by "specify the data coordinates of the viewport"?

Specifying the coordinates, in the original data coordinate system, of the four corners of the viewport (upper right, etc). This is because datoviz takes care of generating the axes, ticks etc, so it needs to know the coordinate system of the user data. So if you do a plot with x spanning [0, 1000] and y spanning [-10, +10], you need to give this information to the visualization library, even though the GPU may only see normalized coordinates in [-1, +1]. This is needed because the GPU uses single precision whereas the CPU can use double precision, and the loss of precision is often not acceptable when dealing with scientific data.

**Comment by rossant** [comment]
@djhoese thanks for the clarification. Non-linear transforms on the GPU may be implemented at some point in Datoviz, depending on user feedback. The way I imagine it is to provide a set of common non-linear transforms that could be selected by the user (eg polar, Earth projections...). But what is tricky is to define interactivity: when you pan with the mouse, how does that affect the non-linear GPU transform? What's the role of the MVP matrices? I don't think this feature will make it into the first 0.1 release because it requires careful planification.

Regarding even more complex transforms with GPU-backed scene graphs, I don't think that would be in the scope of datoviz. Do you have specific use-cases where this would be required (i.e., where arbitrary CPU-backed transforms + simple, predefined linear or non-linear GPU transforms would not be enough)?

If this is required by VisPy, I agree that the underlying GPU visualization backend should then provide a way for the high-level library to specify the GLSL (or SPIR-V) code directly. In this case, I think the best would be for the backend to know nothing about it, and leave the full shader construction to the high-level library. Doing this would bypass all the intermediate layers, for example in Datoviz, the graphics+visuals+panel (subplot)+transform+interactivity+controllers+scene interfaces, and would just rely on the gloo-like interface (vklite+canvas).

**Comment by djhoese** [comment]
Another use of dynamic shaders in VisPy are the "filters": https://github.com/vispy/vispy/blob/master/vispy/visuals/filters/base_filter.py. We have more and more features that these are being used for. For example, adding a texture on top of a Mesh or controlling the alpha value for the entire Visual. Some are simple, some are more complex, but they take advantage of VisPy's "pre" and "post" hooks for shaders (entry points where extra functions can be defined for the shaders). I'm not saying this is needed or that it is the best way to implement it (shadergraph may be a better replacement). If VisPy doesn't have the ability to control things like this then we either need to:

1. Implement the functionality in the lower-level libraries. That may mean a lot more VisPy visuals having to be re-implemented as "primitives" in the lower-level library.
2. Lose the functionality (at least for backends that don't support it)

> But what is tricky is to define interactivity: when you pan with the mouse, how does that affect the non-linear GPU transform? What's the role of the MVP matrices? I don't think this feature will make it into the first 0.1 release because it requires careful planification.

My only experience with non-linear transforms is geographic projection transformations. In this case, you may be doing a non-linear transformation from lon/lat (for example) to polar-stereographic projection space, but the projection space is still linear (X, Y) metered coordinates. So in my own applications this ends up being a chain of transforms: scale/translate -> non-linear projection -> scale/translate, or something similar. I'm not sure the best way to implement this in something like datoviz. Maybe MVP is one option but a callback function for transforms is another? That said, I don't see why pre-defined non-linear transforms wouldn't work **however** you would likely get to the point where enough of these transforms should be put in a separate library (ex. a GLSL 4.x library of geographic transforms) and in that case how would those be "plugged in" to datoviz? What would stop VisPy from using that same interface?

> In this case, I think the best would be for the backend to know nothing about it, and leave the full shader construction to the high-level library. Doing this would bypass all the intermediate layers, for example in Datoviz, the graphics+visuals+panel (subplot)+transform+interactivity+controllers+scene interfaces, and would just rely on the gloo-like interface (vklite+canvas).

Yeah this is a difficult decision. On one hand I think being able to provide a custom shader is extremely useful and I understand that it complicates interfaces that would otherwise be clean and not need a low-level understanding. On the other hand, there are small modifications to shaders that can make them useful in a lot of other use cases. For example, @tlambert03 just started a pull request that uses the ImageVisual as a base class and instead of using the `.r` of the data, he stores a complex number as `RG32F` data and uses the `.r` and `.g` in a custom piece of shader code to determine the resulting color. :thinking: Although I suppose this is a case where this could be done on the CPU if needed.

I think there would also have to be a line drawn (visualization pun not intended) on what kind of custom shaders could be provided. The example of using `.r` and `.g` when the default shader uses only `.r` is a pretty simple change. But wanted to add additional uniforms or buffers is much more complicated and I don't think there is a way to do it without tapping into the low-level library.

Changing topic:

@almarklein The reason I asked the 2D versus 3D question is that we currently have issues with VisPy's ArrowVisual where it was written with the assumption that it would only ever be used in 2D. Users wanted to start using it in 3D land and noticed that it didn't work quite right. We modified it so it didn't ignore the `.z` coordinate and then it worked better. However, users later discovered that if you look perpendicular to the arrow head it suddenly turns into a square or circle due to some math error in the shader (I think). I just want to make sure low-level libraries are designing "2D" visuals with the idea that they might be used in 3D applications.

@rougier Regarding your comment on thinking about the API we want, I think we need to focus on what the properties are that we expect from the low-level libraries. Dynamic shaders is important so that's a separate topic. As far as how/what we pass to the library for other basic properties I think we can assume that VisPy will handle any translation/conversion needed to get the information/data to the low-level library.

`</braindump>`

**Comment by rougier** [comment]
Log / Semi-log axis is a good example of required non-linear transform.

**Comment by rossant** [comment]
I'd say there are two opposite approaches here:

- **The "rigid" approach**: using essentially static shaders, implementing pre-defined visuals, pre-defined features, pre-defined interactivity controllers, implementing specific use-cases on-demand based on user feedback
- **The "flexible" approach**: providing a "plug-and-play" system where shader snippets can be composed together, allowing for an infinite number of combinations and high customizability by power users

So far I chose the rigid approach in Datoviz: it's less powerful, but it's much simpler and it provides perhaps 80% of the use-cases. VisPy follows the flexible approach, but it has a cost in terms of code complexity. Vulkan can hopefully reduce this complexity a bit, but one still has to develop and maintain a whole shader composition system (using either GLSL or SPIR-V).

How could a flexible library (VisPy) be based on a rigid one (datoviz)? I see at least 2 solutions:

- VisPy could only use the vklite and the canvas layers: it would be up to VisPy to construct the shaders, define and implement the visuals and the high-level layers.

- VisPy could reuse the Datoviz visuals, but we would have to think about a way to reuse the shaders and plug custom shader code at some specific points (for example,  `/* */` comment hooks). For the transformations, all Datoviz visuals currently use a built-in GLSL function `vec4 transform(vec3)` that is implemented in a `common.glsl` file #included in all shaders. Perhaps this could be used by VisPy to plug a custom transformation function that implements scene graph transforms.

**Comment by djhoese** [comment]
Good summary. Thanks. @rossant How "all or nothing" is the visual interface in datoviz? By that I mean, how possible would it be for VisPy to write its own datoviz Visuals using vklite and then tie it into the controllers and other higher level features of datoviz? Put another way, is datoviz "rigid" at the high level (visuals/canvas) or at the low-level closer to vklite (does this include graphics)?

I think if we don't take advantage of datoviz's higher level functionality we're not really doing it "right". VisPy can provide some really flexible/dynamic stuff at its own high level with plotting APIs and a SceneCanvas to handle parent/child relationships, but if that comes at the cost of re-implementing Visuals in vklite then we're barely using datoviz. At least that's my understanding. And we wouldn't benefit from as many of the future developments/contributions in datoviz.

I guess my hope was that by making datoviz more "generic", it would not only simplify/modular-ize datoviz but also provide more flexibility to VisPy. I'm not sure I completely understand the transform/common.glsl though. Is that something that is searched for and linked at runtime? Or compile time? If at runtime, could we do some kind of hooking system with multiple includes?

Bottom line: I don't want to make your life hell by requesting all of this stuff of datoviz. Datoviz should do what datoviz wants and what you think is best and most maintainable. I am OK putting everything complex into VisPy as long as datoviz (wgpu/pygfx) can let me access some hooks to customize some things. Even if we can't and we have to limit what VisPy can do to get the better performance of Vulkan/datoviz/WGPU, I'm ok with that too. Looks like I'll have to dust off my C skills and learn some Vulkan to see if I have any other ideas for Datoviz.

It's software right? We can do anything. :fearful:

**Comment by rougier** [comment]
I agree that a rigid approach makes sense and could probably cover most of user's need. However, we have also to think about extension points (even if no extension exists). For example, @rossant how do you plan to implement log axis and if you plan to restrict to linear case, where exactly expert users would need to inject GLSL code?

**Comment by rossant** [comment]
First, I should be clearer and use the proper Datoviz terminology of **graphics**, which is a sort of "low-level visual", compared to "high-level visuals", just called **visuals**. A graphics is really nothing more than a pair of vertex shader and fragment shader, along with some fixed state.

Although Datoviz is "rigid" when it comes to the shaders (all shaders are hard-coded and compiled to SPIR-V at build time), I spent a lot of time ensuring the library itself is modular. Let's consider a graphics, with static, precompiled vertex shader and fragment shader. How to integrate it in a "scene", with multiple subplots, transforms, interactivity? In Datoviz, the chosen approach is to *request all shaders to share the same conventions*, which are currently:

- First binding point of all shaders is a uniform with MVP matrices (model, view, projection)
- Second binding is a uniform with viewport information (offset, size in pixels, etc)
- All shaders must have `#include "common.glsl"` at the top, which is just copying the same GLSL code snippet into all shaders, at build time (works exactly like the `#include` macro in C, this is provided by the official GLSL->SPIR-V compilers)
- The vertex position is always transformed via `gl_Position = transform(pos)`, where `transform()` is a function defined in `common.glsl`, which (currently) just applies the MVP matrices, taken from the first uniform binding.

That's about it. How to use a graphics?

1) If you just have a raw canvas, you have to:

- Create a GPU buffer with the MVP matrices and bind it to the shader
- Create a GPU buffer with the viewport struct and bind it to the shader
- Create a GPU buffer with the vertex buffer and attach it to the shader
- Same for graphics-specific additional binding points (uniform buffers, textures, etc)
- Compute the MVP matrices at every frame and upload them to the MVP buffer
- Upload the viewport struct to the viewport buffer
- Take the user data, normalize it to [-1, +1] by keeping the position data in double precision and casting it to single precision at the very last moment (to avoid loss of precision), compute the vertices (which may involve "baking", triangulation...) and upload them to the vertex buffer

2) If you use the Datoviz scene, you just have to specify the vertex buffer, specify the subplot and controller, and the scene interface will automatically make the appropriate GPU buffers for MVP and viewport, it will update them with the panel position and interactivity following user events (mouse, keyboard, timer...). When the graphics is wrapped into a high-level visual, the vertex data computation is done automatically on the CPU in the transformation (panel-level) and baking (visual-level) functions.

Hopefully you should now have a better idea of how Datoviz works at the moment.

To me, the most important issue in all of this is **shader composition**: how to allow for custom shader extensibility so that you don't have to manually copy-paste existing shaders to make small changes in them. The good thing with the current Datoviz architecture is that it's all relatively independent from the shaders: the only things Datoviz assumes in the shaders are the first binding points (MVP and viewport), and the `gl_Position = transform(pos)` function. These are just conventions: the builtin graphics shaders are *just normal GLSL shaders* that could be used with any other OpenGL/Vulkan backend. There is no custom syntax (like in VisPy, I think) that makes them non directly usable elsewhere. I would like to keep it this way.

What we have to determine now is how to make the existing Datoviz shaders, which are 100% pure GLSL, reusable and extensible by a higher-level library based on Datoviz. That could be done either (1) at the GLSL level or (2) at the SPIR-V level. Perhaps the first step would be for someone to go through a few existing shaders in Datoviz (marker, image, path..?), and in each of them, determine where it would make sense to have custom "hooks". I would be happy to implement any extensibility system in the included shaders *that leaves them 100% pure GLSL*.

The most important hook would be transform, and it's already all implemented in `transform()`. So a library that would be based on Datoviz could perhaps load the included shaders and swap the builtin `transform()` with a custom one, either in GLSL or in SPIR-V bytecode.

**Comment by rossant** [comment]
> I agree that a rigid approach makes sense and could probably cover most of user's need. However, we have also to think about extension points (even if no extension exists). For example, @rossant how do you plan to implement log axis and if you plan to restrict to linear case, where exactly expert users would need to inject GLSL code?

The poor man's approach is to compute the log on the CPU, upload the log-transformed data on the GPU, and use a special log axes visual. As a first approach, panning and zooming could be done the normal way, with linear transforms only. It's not perfect, but it's a start.

For custom GLSL code injection, as I said in my other comment above, one way would be to parse the call to `transform()` and replace it with a custom transformation function either in GLSL or SPIR-V. I don't know SPIR-V well enough to understand how feasible it would be to replace the existing transform function with another one directly in the bytecode.

**Comment by djhoese** [comment]
Great! I think I actually understood most of this :wink: 

I appreciate your strong stance on keeping the shaders pure GLSL and I agree. I have like 3 other things above this on my immediate priority list, but this is more fun so I'll probably get to it this week. I'll look at datoviz, try to get some examples running locally, and see if I can bend it to my will :smiling_imp:

**Comment by rossant** [comment]
I had a quick look at the disassembled SPIR-V of a Datoviz shader. In principle you have access to the bytecode of all GLSL functions, for example one polymorphic version of `transform()` looks like this:

``` 
%transform_vf3_ = OpFunction %v4float None %25
      %pos_0    = OpFunctionParameter %_ptr_Function_v3float
                  ....
                  OpReturnValue %232
                  OpFunctionEnd
```

and it is called like this:

```
%243 = OpFunctionCall %v4float %transform_vf3_ %param_3
```

So if we had access to a good API for SPIR-V bytecode manipulation, we could relatively easily swap dynamically Datoviz functions with user-provided custom functions with the same signature.

For example, we could imagine an API like this:

```c
void dvz_shader_hook(
    DvzShader* shader, DvzShaderHook hook, VkDeviceSize size, const uint32_t* spirv_bytecode);
```

where:

* `shader` represents a given builtin vertex or fragment shader of a given graphics
* `hook`: an enum indicating which hook to use, for example `DVZ_SHADER_HOOK_TRANSFORM` if one wants to swap the builtin `transform()` function with a custom one
* `size`: the number of bytes of the buffer containing the SPIR-V bytecode
* `spirv_bytecode`: a buffer with the SPIR-V bytecode

Then it would be up to the high-level library to define a custom transform (or other) function with the same signature, in GLSL, Python, or any language, compile it to SPIR-V, and pass the compiled bytecode to Datoviz.

**Comment by almarklein** [comment]
> [...] and it makes the entire system GLSL-dependent, whereas the whole point of SPIR-V is to allow shaders to be written in any language.

This is a good point. From my point of view, it's possible that pygfx will eventually use WGSL shaders by default. But perhaps we'll stick to GLSL initially. In short, I consider the used shader language an implementation detail for now.

Modifying SpirV sounds interesting ... and also like something that can become complex quickly. E.g. introducing a new variable may require changes in the "header" as well to define it.

Perhaps we should consider that any shader composition is part of the graphics backends and `vispy.visuals` will just have to work with whatever functionality is offered. Need a polar plot? -> implement it in datovoz or pygfx, and then expose in Vispy.

@rossant how do you deal with variations in shaders, e.g. for showing textures with different formats or e.g. grayscale/rgb? Do you have multiple shaders that are nearly similar, or do you use `#include` to share code? In Pygfx we're also considering e.g. jinja2 for templating, so we have some degree of applying variations and introduce code at specific parts, but with a tool that is simple/predictable/known.

**Comment by rossant** [comment]
> Modifying SpirV sounds interesting ... and also like something that can become complex quickly. E.g. introducing a new variable may require changes in the "header" as well to define it.

The advantage of working only with *functions* is that modifying code dynamically might be simpler (although they are not "pure" in the sense that they may use global constants, uniforms, etc.).

> @rossant how do you deal with variations in shaders, e.g. for showing textures with different formats or e.g. grayscale/rgb? Do you have multiple shaders that are nearly similar, or do you use `#include` to share code? In Pygfx we're also considering e.g. jinja2 for templating, so we have some degree of applying variations and introduce code at specific parts, but with a tool that is simple/predictable/known.

I don't for now. For textures, the image format is not specified in the shader, and not even in a graphics' implementation, only when *using* it. One specifies the texture format when creating a texture, and binds it to the graphics.

However, what *is* hard-coded in the shader is the number of texture dimensions (1D, 2D, or 3D), and whether to use the first component only (`.r`) or more. For the latter, I never had the use-case yet. If I do, I see at least 2 solutions.

First solution (simple but not ideal in terms of performance because of the conditional branch) is to add a parameter field in the uniform param struct, and add a dynamic `if/else` in the shader:

```glsl
float value = 0;
if (params.n_components == 1)
    value = texture(tex, uv).r;
else if (params.n_components == 2)
    value = get_value(texture(tex, uv).rg);
```

Second solution is to use macros, and update my SPIR-V CMake-based build system to generate the different variants at build-time (you can pass macro values when compiling the shaders with something like `-DTEXTURE_RG=1`):

```glsl
#ifdef TEXTURE_RG
float value = get_value(texture(tex, uv).rg);
#else 
float value = texture(tex, uv).r;
#endif
```

If the variant is about attributes/uniforms/etc, only the second solution would work.

**Comment by queenish001** [comment]
@djhoese

**Comment by djhoese** [comment]
We'll keep this open. This is an ongoing effort and discussion.
